web/impact/impact/v1/views/user_program_confirmation_detail_view.py

```diff
@@ -0,0 +1,107 @@
+from django.contrib.auth import get_user_model
+from django.db.models import F
+from django.template import loader
+from rest_framework.exceptions import ParseError
+from rest_framework.response import Response
+from accelerator.models import (
+    UserRole,
+    Program,
+    ProgramRoleGrant,
+    ProgramRole
+)
+from impact.minimal_email_handler import MinimalEmailHandler
+from impact.permissions.v1_api_permissions import UserDetailViewPermission
+from impact.v1.helpers import UserProgramConfirmationHelper
+from impact.v1.views.user_detail_view import (
+    BaseDetailView,
+    NO_USER_ERROR
+)
+
+User = get_user_model()
+
+SUBJECT = '[MassChallenge] Confirmed: Your Mentorship Commitment'
+INVALID_INPUT_ERROR = "'{}' must be a list of numbers."
+
+
+def extract_values(request, key):
+    try:
+        return [int(i) for i in request.data.get(key, [])]
+    except ValueError:
+        raise ParseError(INVALID_INPUT_ERROR.format(key))
+
+
+class UserProgramConfirmationDetailView(BaseDetailView):
+    view_name = "user_program_confirmation_detail"
+    helper_class = UserProgramConfirmationHelper
+    permission_classes = (UserDetailViewPermission,)
+
+    def __init__(self, *args, **kwargs):
+        self.user = None
+        super().__init__(*args, **kwargs)
+
+    def add_program_role_grants(self, program_ids=None, user_role=None):
+        programs = Program.objects.filter(pk__in=program_ids)
+        program_roles = dict(ProgramRole.objects.filter(
+            user_role__name=user_role
+        ).order_by('-id').values_list('program_id', 'id'))
+        program_role_grants = [
+            ProgramRoleGrant(
+                person=self.user,
+                program_role_id=program_roles[program.id]
+            )
+            for program in programs
+        ]
+        ProgramRoleGrant.objects.bulk_create(
+            program_role_grants,
+            ignore_conflicts=True
+        )
+
+    def delete_program_role_grants(self, program_ids=None, user_role=None):
+        self.user.programrolegrant_set.filter(
+            program_role__program__in=program_ids,
+            program_role__user_role__name=user_role
+        ).delete()
+
+    def update_user_confirmation(self, delete_roles, add_roles, program_ids):
+        self.delete_program_role_grants(program_ids, delete_roles)
+        self.add_program_role_grants(program_ids, add_roles)
+
+    def post(self, request, pk):
```

Comment:
POST requests don't usually take args, do they? Shouldn't the user id be passed in as part of POST.data?

Reply:
the `pk` is there to facilitate GET on `/api/v1/user/<pk>/program_confirmation/`.
I used it since the required user `pk` existed as an arg and getting the same `pk`  from POST.data seemed redundant. Do you suggest I switch to getting it from POST.data?

Reply:
Clarified in slack convo.